src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/LazyNullable.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A lazy version of <see cref="Nullable{T}"/> which uses the same space as a <see cref="Nullable{T}"/>.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+internal struct LazyNullable<T>
```

Comment:
💡 Consider renaming this `SingleInitNullable<T>` to distinguish it from `Lazy<T>` (the latter provides the initialization function to the constructor, where this new type doesn't contain any information about how to initialize itself)

Reply:
sgtm.